gapis/replay/export_replay.go

```diff
@@ -86,43 +78,142 @@ func ExportReplay(ctx context.Context, pCapture *path.Capture, pDevice *path.Dev
 
 	b := builder.New(replayABI.MemoryLayout)
 
-	_, ranges, err := initialcmds.InitialCommands(ctx, pCapture)
+	_, ranges, err := initialcmds.InitialCommands(ctx, capturePath)
+
+	out := &adapter{
+		state:   c.NewUninitializedState(ctx, ranges),
+		builder: b,
+	}
 
 	generatorReplayTimer.Time(func() {
-		err = generator.Replay(
+		ctx := status.Start(ctx, "Generate")
+		defer status.Finish(ctx)
+
+		err = m.key.generator.Replay(
 			ctx,
-			Intent{pDevice, pCapture},
-			Config(&struct{}{}),
-			[]RequestAndResult{{
-				Request: Request(generator.(interface{ ExportReplayRequest() Request }).ExportReplayRequest()),
-				Result:  func(val interface{}, err error) {},
-			}},
+			intent,
+			m.key.config,
+			m.requests,
 			d.Instance(),
 			c,
-			&adapter{
-				state:   c.NewUninitializedState(ctx, ranges),
-				builder: b,
-			})
+			out)
 	})
 
 	if err != nil {
-		return log.Err(ctx, err, "Replay returned error")
+		return nil, log.Err(ctx, err, "Replay returned error")
+	}
+
+	if config.DebugReplay {
+		log.I(ctx, "Building payload...")
 	}
 
 	var payload gapir.Payload
-	var handlePost builder.PostDataHandler
-	var handleNotification builder.NotificationHandler
-	builderBuildTimer.Time(func() { payload, handlePost, handleNotification, err = b.Build(ctx) })
+	builderBuildTimer.Time(func() { payload, _, _, err = b.Build(ctx) })
 	if err != nil {
-		return log.Err(ctx, err, "Failed to build replay payload")
+		return nil, log.Err(ctx, err, "Failed to build replay payload")
+	}
+	return &payload, nil
+}
+
+// NOTE: Only works for query that does not require result to be passed back.
+func (m *exportManager) Replay(
+	ctx context.Context,
+	intent Intent,
+	cfg Config,
+	req Request,
+	generator Generator,
+	hints *service.UsageHints) (val interface{}, err error) {
+
+	key := &batchKey{
+		capture:   intent.Capture.ID.ID(),
+		device:    intent.Device.ID.ID(),
+		config:    cfg,
+		generator: generator,
+	}
+
+	if m.key == nil {
+		m.key = key
+	}
+
+	if *key != *m.key {
+		return nil, log.Errf(ctx, nil, "Can not export trace with incompatible requests")
+	}
+
+	type res struct {
+		val interface{}
+		err error
+	}
+	out := make(chan res, 1)
+	m.requests = append(m.requests, RequestAndResult{
+		Request: req,
+		Result:  Result(func(val interface{}, err error) { out <- res{val, err} }),
+	})
+	return nil, nil
+}
+
+type ExportReplayFramebufferRequest struct {
+	Request          *service.GetFramebufferAttachmentRequest
+	Width, Height    uint32
+	FramebufferIndex uint32
+}
+
+// ExportReplay write replay commands and assets to path.
+func ExportReplay(ctx context.Context, pCapture *path.Capture, pDevice *path.Device, outDir string, fbs []ExportReplayFramebufferRequest) error {
```

Comment:
Instead of having the caller build and pass in `ExportReplayFramebufferRequest`s, how about:

```go

exporter := replay.Exporter() // Returns an instance of your new replay.Manager 

// Use the existing API methods for building replays. The code in gapis/resolve/framebuffer_attachment_data.go could be changed to support both cases.
query, ok := api.(replay.QueryFramebufferAttachment)

for _, fbr := range FramebufferAttachmentRequest {
	// Call the existing API's QueryXXX functions to add replay requests to the exporter.
	_, err := query.QueryFramebufferAttachment(ctx, intent, exporter, ...)
}

// Call mgr.generate, spit out the files...
err := exporter.Export(outDir)
```

Your `Exporter.Export` function then doesn't have to do all the translation to Query calls itself, and you'll automatically support any future Query functions.
Note in `Export` you'll probably want to check that all requests share the same `replay.Config`, otherwise you're going to be building multiple replay batches (which maybe something it would helpful to handle in the future?).

Reply:
That's a much cleaner solution. Thanks.